libs/api/libbpf_program.cpp

```diff
@@ -0,0 +1,236 @@
+        return nullptr;
+    }
+
+    // TODO: use ifindex
+    UNREFERENCED_PARAMETER(ifindex);
+
+    ebpf_handle_t link_handle;
+    uint32_t result = ebpf_api_link_program(program->handle, EBPF_ATTACH_TYPE_XDP, &link_handle);
+    if (result) {
+        return nullptr;
+    }
+
+    // The bpf_link structure is opaque so we can use the link handle directly.
+    return (struct bpf_link*)link_handle;
+}
+
+struct bpf_program*
+bpf_program__next(struct bpf_program* previous, const struct bpf_object* object)
+{
+    return ebpf_program_next(previous, object);
+}
+
+struct bpf_program*
+bpf_program__prev(struct bpf_program* next, const struct bpf_object* object)
+{
+    return ebpf_program_previous(next, object);
+}
+
+int
+bpf_program__unpin(struct bpf_program* program, const char* path)
+{
+    char fullpath[MAX_PATH];
+    int length = snprintf(fullpath, sizeof(fullpath), "%s/%s", path, program->program_name);
+    if (length < 0) {
+        return -EINVAL;
+    } else if (length > sizeof(fullpath)) {
+        return -ENAMETOOLONG;
+    }
+
+    uint32_t result = ebpf_api_unpin_object((const uint8_t*)fullpath, (uint32_t)strlen(fullpath));
+    return (int)result;
+}
+
+static void
+_unpin_previous_programs(_In_ ebpf_program_t* program, _In_ struct bpf_object* object, const char* path)
+{
+    for (program = bpf_program__prev(program, object); program; program = bpf_program__prev(program, object)) {
+        bpf_program__unpin(program, path);
+    }
+}
+
+int
+bpf_program__pin(struct bpf_program* program, const char* path)
+{
+    char fullpath[MAX_PATH];
+    int result;
+
+    int length = snprintf(fullpath, sizeof(fullpath), "%s/%s", path, program->program_name);
+    if (length < 0) {
+        return libbpf_err(-EINVAL);
+    }
+
+    result = ebpf_api_pin_object(program->handle, (const uint8_t*)fullpath, (uint32_t)strlen(fullpath));
+    if (result) {
+        return libbpf_err(result);
+    }
+
+    return 0;
+}
+
+int
+bpf_object__pin_programs(struct bpf_object* object, const char* path)
+{
+    struct bpf_program* program;
+
+    if (!object) {
+        return libbpf_err(-ENOENT);
+    }
+
+    bpf_object__for_each_program(program, object)
+    {
+        int result = bpf_program__pin(program, path);
+        if (result) {
+            _unpin_previous_programs(program, object, path);
+            return result;
+            ;
+        }
+    }
+
+    return 0;
+}
+
+int
+bpf_object__unpin_programs(struct bpf_object* object, const char* path)
+{
+    struct bpf_program* program;
+
+    if (!object) {
+        return libbpf_err(-ENOENT);
+    }
+
+    bpf_object__for_each_program(program, object)
+    {
+        int result = bpf_program__unpin(program, path);
+        if (result) {
+            return libbpf_err(result);
+        }
+    }
+
+    return 0;
+}
+
+int
+bpf_link__destroy(struct bpf_link* link)
+{
+    ebpf_handle_t link_handle = (ebpf_handle_t)link;
+    uint32_t result = ebpf_api_close_handle(link_handle);
```

Comment:
This bpf_link__destroy() also needs to call DETACH API before closing fd. I think we can add a TODO here to track this change once the DETACH changes PR is completed?

Reply:
Added TODO as suggested